Add click handler to OverviewView that centers the detail view

andreasg123 opened the pull request:
This addresses #330. I don't have an easy way to test this but a very similar implementation with subclasses worked in my app. I only tried it with one image but there the combination of `zoom` and `scale` properly transformed the coordinates such that the mouse was in the center of the red bounding rectangle after a click.

andreasg123 commented:
@ilan-gold, the way your repo is set up, I can't request a review.

ilan-gold commented:
@andreasg123 Ok, thanks for the heads up.

ilan-gold commented on src/views/OverviewView.js:
```diff
@@ -1,25 +1,53 @@
+    const scaleFactor = 1 / (2 ** zoom * scale);
+    x *= scaleFactor;
+    y *= scaleFactor;
+    if (this.onViewStateChange) {
+      this.onViewStateChange({ viewState: { Overview: { target: [x, y, 0] } } });
+    }
```
```suggestion
    const { id } = this;
    if (this.onViewStateChange) {
      this.onViewStateChange({
        viewId: id,
        viewState: { id, target: [x, y, 0] }
      });
    }
```

I think this is what we want, modulo prettier?  Did what you have work with Avivator?  It might have been relying on some of the old overriding code you had.

andreasg123 replied:
I'm not sure if that would work because `viewState.target` has a meaning for `OverviewView`. I'll test it in my own implementation. However, I just noticed a mistake that I made in string replace, instead of `Overview`, that should have been `navigation`.

andreasg123 replied:
It works without `navigation` so that I'll remove that.

Why do you want to add `id` and `viewId`? deck.gl's `ViewManager` automatically adds `viewId`: https://github.com/visgl/deck.gl/blob/master/modules/core/src/lib/view-manager.js#L240
`VivViewer._onViewStateChange` adds that to `viewState` as `id`.

ilan-gold replied:
@andreasg123 I didn't know that about `id` so you could remove that.  I think the `id` should be `overview` not `navigation` since that is the internally used `id`.  Your other change wrt. `navigation` had `viewState.navigation` but it seems like you're saying we should make `viewState.id` to be `navigation` which is different.

ilan-gold commented on src/views/DetailView.js:
```diff
@@ -43,4 +43,20 @@ export default class DetailView extends VivView {
+   */
+  filterViewState({ viewState, currentViewState }) {
+    if (viewState.id === 'overview') {
+      const target = viewState.navigation?.target;
```
```suggestion
      const { target } = viewState;
```
Going along with the other requested change, but I may have this wrong again.  Also could we put the `id`s in a `constants.js` file now that they are being cross-referenced?  I think just `DetailView`'s id (which is `detail`) and `OverviewView`'s (which is `overview`) should be there.  Thanks!

ilan-gold commented on src/views/OverviewView.js:
```diff
@@ -1,25 +1,53 @@
+  }
+
+  handleEvent(event) {
+    if (event.type !== "click") {
```
I think this would fail prettier with the double-quotes - could you enable GH actions in your fork repo while we resolve the fact that forking PR's don't run actions (#333)?

andreasg123 replied:
I have a quotes rule in my `.eslintrc.json`: `"quotes": [2, "double", "avoid-escape"]`. May I add that rule, just with single quotes, to yours?

ilan-gold replied:
@andreasg123 Do you have an `eslint` file overriding ours?  I am a little confused since our `.eslintrc.yml` extends prettier.

andreasg123 replied:
No, I talked about a different repo. This was just a suggestion how to enforce single quotes but just using prettier is fine

ilan-gold reviewed (changes requested):
Could you check this against a local Avivator demo from this repo? Thanks!  Seems to work well with the suggested changes so I think we are close here.  I had never used `Controller` before - deck.gl is so full of great features.

andreasg123 commented:
I successfully tested the updated version in Avivator.

andreasg123 commented:
After looking at the output of GH Actions, I added the Prettier changes just to the files that I changed. I don't necessarily agree with them, e.g.:
```
v => !v
(v) => !v
```

Let me know if you want to keep those changes or whether I should revert to the previous commit.

ilan-gold commented:
I think we need the prettier changes to pass the gh actions - if you use `vscode` you should be able to turn on prettier to run after every save, or you could run `npm run format:write`.

ilan-gold commented:
If you want to open a separate PR about our prettier settings, that is also fine.

andreasg123 commented on src/viewers/PictureInPictureViewer.js:
```diff
@@ -44,15 +51,16 @@ const PictureInPictureViewer = props => {
     initialViewState: detailViewState,
     height,
-    width
+    width,
```
This comma was added by `prettier --write` and has nothing to do with my changes.

andreasg123 commented on src/viewers/PictureInPictureViewer.js:
```diff
@@ -65,18 +73,19 @@ const PictureInPictureViewer = props => {
     lensRadius,
     lensBorderColor,
-    lensBorderRadius
+    lensBorderRadius,
```
Same for this comma.

andreasg123 commented:
If I run `prettier --check` on `master`, I get this:
```
$ prettier --check src/**/*.js avivator/src/**/*.js tests/**/*.js
Checking formatting...
[warn] src/layers/BitmapLayer.js
[warn] src/layers/ImageLayer.js
[warn] src/layers/OverviewLayer.js
[warn] src/layers/ScaleBarLayer.js
[warn] src/layers/utils.js
[warn] src/loaders/OMETiffLoader.js
[warn] src/loaders/Pool.js
[warn] src/loaders/index.js
[warn] src/loaders/omeXML.js
[warn] src/loaders/utils.js
[warn] src/loaders/viv.worker.js
[warn] src/loaders/zarrLoader.js
[warn] src/viewers/PictureInPictureViewer.js
[warn] src/viewers/SideBySideViewer.js
[warn] src/viewers/VivViewer.js
[warn] src/views/DetailView.js
[warn] src/views/OverviewView.js
[warn] src/views/SideBySideView.js
[warn] src/views/VivView.js
[warn] src/views/utils.js
[warn] avivator/src/components/ChannelController.js
[warn] avivator/src/components/ChannelOptions.js
[warn] avivator/src/components/ColorPalette.js
[warn] avivator/src/components/ColormapSelect.js
[warn] avivator/src/components/Dropzone.js
[warn] avivator/src/components/GlobalSelectionSlider.js
[warn] avivator/src/components/LensSelect.js
[warn] avivator/src/components/Menu.js
[warn] avivator/src/components/MenuTitle.js
[warn] tests/loaders/OMETiffLoader.spec.js
[warn] tests/loaders/utils.spec.js
[warn] tests/loaders/zarrLoader.spec.js
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
```

My point about GH Actions is that they are not that helpful if the repo already fails without any changes. I would suggest that you run `prettier --write` after you settle on your prettier configuration so that there is a good starting point.
